### DISCUSSION TITLE
Avalonia Port of WPF DataGrid

### DISCUSSION THREAD

**Discussion by robloo** [AvaloniaUI/Avalonia#14259] [Ideas]
The current version of the Avalonia DataGrid is based off the old Silverlight version. This was the best way to do it back then as no other source code was available with appropriate licensing. However, the Silverlight version has a number of issues compared to the WPF version which is more complete and tested. Instead of trying to fix the Silverlight implementation it is probably better to re-base the code on the WPF version. Then bug fixes can be applied in the WPF repository as well as Avalonia benefiting everyone.

https://github.com/dotnet/wpf/blob/master/src/Microsoft.DotNet.Wpf/src/PresentationFramework/System/Windows/Controls/DataGrid.cs

**Comment by Shadow4walker** [comment]
I would love to see that too the current version have so many bugs which is sucks!

**Comment by kekekeks** [comment]
We'll probably need to implement https://github.com/AvaloniaUI/Avalonia/issues/2769 and triggers first. That would radically simplify the porting process

**Comment by maxkatz6** [comment]
Also there is DataTable from WindowsCommunityToolkit for UWP with open sourced code - https://docs.microsoft.com/en-us/windows/communitytoolkit/controls/datagrid
But I didn't use it and not sure, if it is better or worse than wpf version.

**Comment by robloo** [comment]
@Tirraon The WindowsCommunityToolkit is also using the Silverlight base for their DataGrid implementation. There are many issues with it as well and most have gone unfixed. The Silverlight version was never really at the level of the WPF one.

**Comment by jmacato** [comment]
I do have an experimental DataGrid like control on my branch `https://github.com/AvaloniaUI/Avalonia/tree/xdatagrid-experiment` that doesn't use any prior code from WPF/Silverlight and is powered by `ItemsRepeater`. Still a WIP 

![image](https://user-images.githubusercontent.com/16554748/64906242-91fc6380-d716-11e9-812e-7acdd487bc8a.png)

**Comment by robloo** [comment]
@jmacato Having a light-weight alternative is definitely useful in some cases. But Avalonia benefits immensely from keeping the full featured DataGrid control as well.

**Comment by Gillibald** [comment]
I don't think that anyone of the Avalonia devs will have time soon to work on a rich DataGrid implementation. Contributions by the community are always welcome.

**Comment by jmacato** [comment]
It is an extremely complex control (and frankly I hate its code and it'd be unmaintainable given the lack of resources and time on our part). But if someone can contribute with that then that'd be great.

**Comment by robloo** [comment]
All good points and I'm not saying this be attempted right now. I also realize how large DataGrid is. 

It's possible a refactor would be allowed in the actual WPF repo and then once that's done the code would be more easily ported. I might be willing to take a stab at that myself but I think it's going to take at least until the 0.10 release for the Avalonia feature set to be ready to make porting controls like this reasonable.

**Comment by sdoroff** [comment]
Is anyone currently working on this? I might give it a go.

**Comment by robloo** [comment]
@sdoroff I'm not working on it and have no plans to for the foreseeable future (many other things take priority and the existing one is good-enough for now). If you want to work on it I think that would be great!

**Comment by dlukach** [comment]
Any updates on this? Does anyone want to collab on implementing a datagrid? I'm in desperate need to have a datagrid that supports touch gesturing which the silverlight ported one does not have. Don't mind working with someone who has more experience designing avalonia/wpf controls

**Comment by maxkatz6** [comment]
CC @grokys
